drivers/wifi/infineon/airoc_whd_hal.c

```diff
@@ -0,0 +1,474 @@
+				    &buffer_if_default, &netif_if_default);
+	if (result == CY_RSLT_SUCCESS) {
+		result = whd_bus_sdio_attach(data->whd_drv, &whd_sdio_config,
+					     (whd_sdio_t)&data->card);
+
+		if (result == CY_RSLT_SUCCESS) {
+			result = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (result != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
+		}
+	}
+	return result;
+}
+
+/*
+ * Implement SDIO CMD52/53 wrappers
+ */
+whd_result_t whd_bus_sdio_cmd52(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+				whd_bus_function_t function, uint32_t address, uint8_t value,
+				sdio_response_needed_t response_expected, uint8_t *response)
+{
+	whd_result_t result;
+	struct sd_card *sd = (struct sd_card *)whd_driver->bus_priv->sdio_obj;
+	sdio_cmd_argument_t arg = {0};
+	struct sdhc_command cmd = {.opcode = SDIO_RW_DIRECT,
+				   .response_type = SD_RSP_TYPE_R5,
+				   .timeout_ms = CONFIG_SD_CMD_TIMEOUT};
+
+	arg.cmd52.function_number = (uint32_t)(function & BUS_FUNCTION_MASK);
+	arg.cmd52.register_address = (uint32_t)(address & 0x00001ffff);
+	arg.cmd52.rw_flag = (uint32_t)((direction == BUS_WRITE) ? 1 : 0);
+	arg.cmd52.write_data = value;
+	cmd.arg = arg.value;
+	WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd52);
+
+	result = sdhc_request(sd->sdhc, &cmd, NULL);
+
+	WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(whd_driver->bus_priv, (result != WHD_SUCCESS),
+						     cmd52_fail);
+	if (response != NULL) {
+		*response = cmd.response[0U] & SDIO_DIRECT_CMD_DATA_MASK;
+	}
+
+	/* Possibly device might not respond to this cmd. So, don't check return value here */
+	if ((result != WHD_SUCCESS) && (address == SDIO_SLEEP_CSR)) {
+		return result;
+	}
+
+	CHECK_RETURN(result);
+	return WHD_SUCCESS;
+}
+
+whd_result_t whd_bus_sdio_cmd53(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+				whd_bus_function_t function, sdio_transfer_mode_t mode,
+				uint32_t address, uint16_t data_size, uint8_t *data,
+				sdio_response_needed_t response_expected, uint32_t *response)
+{
+	whd_result_t result;
+	struct sd_card *sd = (struct sd_card *)whd_driver->bus_priv->sdio_obj;
+	sdio_cmd_argument_t arg = {0};
+
+	if (direction == BUS_WRITE) {
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_write);
+	}
+
+	if (direction == BUS_READ) {
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_read);
+	}
+
+	arg.cmd53.function_number = (uint32_t)(function & BUS_FUNCTION_MASK);
+	arg.cmd53.register_address = (uint32_t)(address & BIT_MASK(17));
+	arg.cmd53.op_code = (uint32_t)1;
+	arg.cmd53.rw_flag = (uint32_t)((direction == BUS_WRITE) ? 1 : 0);
+
+	if (mode == SDIO_BYTE_MODE) {
+		__ASSERT(data_size <= (uint16_t)512,
+			 '"%s...": whd_bus_sdio_cmd53: data_size > 512 for byte mode', __func__);
+		arg.cmd53.count = (uint32_t)(data_size & 0x1FF);
+
+	} else {
+		arg.cmd53.count = (uint32_t)((data_size / (uint16_t)SDIO_64B_BLOCK) & BIT_MASK(9));
```

Comment:
Again, fine to use the "magic number" 9 here, but a comment on why its the right number should be added.

Reply:
updated